fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/api/transfer/v1/transaction/Transaction.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.transfer.v1.transaction;
+
+import net.fabricmc.fabric.impl.transfer.transaction.TransactionImpl;
+
+/**
+ * A global operation where {@linkplain Participant participants} guarantee atomicity:
+ * either the whole operation succeeds, or it is completely cancelled.
+ * In that case, we say that it is aborted.
+ *
+ * <p>It is possible to open a transaction when anoter transaction is active, using {@link #openNested}.
+ * In that case, we say that the new transaction is nested in the outer transaction.
+ * Transaction state can be viewed as a stack.
+ * Nested transactions can be committed or rolled back like a regular transaction,
+ * but if a transaction is rolled back all its nested transactions will be rolled back as well,
+ * even if they were committed
+ *
+ * <p>In practice, this means that when a participant enlists itself in a transaction,
+ * the transaction manager ensures that it's enlisted in all the transactions in the stack.
+ * It is guaranteed that {@link Participant#onEnlist} will be called for a parent transaction
+ * before it's called for a child transaction.
+ *
+ * <p>{@link Participant#onClose} will be called for every closed transaction,
+ * but a committed nested transaction may be aborted later.
+ * As such, it is better to defer irreversible success until {@link Participant#onFinalCommit} is called,
+ * which will only be called on success for the root transaction.
+ *
+ * <p>Only one outermost transaction can be open at any given time.
+ * Attempts to open a new outer transaction from the thread of the active transaction will throw a RuntimeException.
+ * Attempts to open a new outer transaction from another thread will block until the active transaction is closed.
+ * The server thread is scheduled before the other threads when possible.
+ * Still, try to keep the transaction work on the other threads minimal.
+ */
+public interface Transaction extends AutoCloseable {
```

Comment:
:iapprove:

Reply:
Resolving because this PR thread is too big already, and this is just the beginning. 😅